File: twoSum.ts
function twoSum(nums: number[], target: number): number[] | null {
  const seen = new Map<number, number>();
  for (let i = 0; i < nums.length; i++) {
    const complement = target - nums[i];
    if (seen.has(complement)) {
      return [seen.get(complement)!, i];
    }
    seen.set(nums[i], i);
  }
  return null;
}

// Quick self-test
const tests: [number[], number, number[] | null][] = [
  [[], 9, null],
  [[7], 14, null],
  [[1, 3, 5, 7], 20, null],
  [[-4, -1, 0, 3, 5], -1, [0, 3]],
  [[3, 2, 3], 6, [0, 2]],
  [[1, 4, 2, 3, 5], 6, [1, 2]],
  [[0, 0, 0], 0, [0, 1]],
];

tests.forEach(([nums, target, expected], i) => {
  const result = twoSum(nums, target);
  const pass = JSON.stringify(result) === JSON.stringify(expected);
  console.log(`Test ${i + 1}: ${pass ? "PASS" : "FAIL"} → ${result}`);
});

File: two_sum.py
def two_sum(nums, target):
    seen = {}
    for i, num in enumerate(nums):
        complement = target - num
        if complement in seen:
            return [seen[complement], i]
        seen[num] = i
    return None


def run_tests():
    tests = [
        ([], 9, None),
        ([7], 14, None),
        ([1, 3, 5, 7], 20, None),
        ([-4, -1, 0, 3, 5], -1, [0, 3]),
        ([3, 2, 3], 6, [0, 2]),
        ([1, 4, 2, 3, 5], 6, [1, 2]),
        ([2**31 - 2, 1, -2**31], 2**31 - 1, [0, 1]),
        ([0, 0, 0], 0, [0, 1]),
    ]

    for idx, (nums, target, expected) in enumerate(tests, 1):
        result = two_sum(nums, target)
        status = "PASS" if result == expected else "FAIL"
        print(f"Test {idx}: {status} | nums={nums} target={target} → {result}")

if __name__ == "__main__":
    run_tests()
